File: frontend/flutter_app/lib/models/usuario.dart
/// Representa un usuario del sistema OhMyFreezer.
///
/// [esJefeCocina] determina qué funcionalidades están disponibles:
/// - `true`  → puede crear/editar recetas y recibe alertas de stock.
/// - `false` → solo puede consultar recetas y elaborarlas.
class Usuario {
  final int id;
  final String username;
  final bool esJefeCocina;
  final String fechaRegistro; // ISO-8601
  final String? email;

  const Usuario({
    required this.id,
    required this.username,
    required this.esJefeCocina,
    required this.fechaRegistro,
    this.email,
  });

  factory Usuario.fromJson(Map<String, dynamic> json) => Usuario(
    id: json['id'] as int,
    username: json['username'] as String,
    esJefeCocina: json['esJefeCocina'] as bool,
    fechaRegistro: json['fechaRegistro'] as String,
    email: json['email'] as String?,
  );

  Map<String, dynamic> toJson() => {
    'id': id,
    'username': username,
    'esJefeCocina': esJefeCocina,
    'fechaRegistro': fechaRegistro,
    'email': email,
  };

  Usuario copyWith({
    int? id,
    String? username,
    bool? esJefeCocina,
    String? fechaRegistro,
    String? email,
  }) => Usuario(
    id: id ?? this.id,
    username: username ?? this.username,
    esJefeCocina: esJefeCocina ?? this.esJefeCocina,
    fechaRegistro: fechaRegistro ?? this.fechaRegistro,
    email: email ?? this.email,
  );
}

File: backend/src/main/java/com/ares/backend/entity/Usuario.java
package com.ares.backend.entity;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.Filter;

import java.time.LocalDateTime;

/**
 * Entidad que representa un usuario del sistema OhMyFreezer.
 * Puede ser un cocinero normal o un jefe de cocina con permisos especiales.
 *
 * @author Ares
 * @version 1.0
 */
@Entity
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@ToString(exclude = "password")
@Table(name = "usuarios", uniqueConstraints = @UniqueConstraint(
        name = "uk_usuarios_negocio_username", columnNames = {"negocio_id", "username"}))
@Filter(name = "negocioFilter", condition = "negocio_id = :negocioId")
public class Usuario {

    /**
     * Identificador único del usuario.
     */
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    /**
     * Negocio (tenant) al que pertenece este usuario. La unicidad del
     * username ya no es global: se recompone como UNIQUE(negocio_id, username).
     * Tanto {@code UsuarioService.registrar()} como
     * {@code UsuarioService.crearEmpleado()} setean siempre este campo antes
     * de persistir, por lo que el mapeo JPA lo refleja como obligatorio
     * (consistente con la constraint NOT NULL de la migración V2 en BD).
     */
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "negocio_id", nullable = false)
    private Negocio negocio;

    /**
     * Nombre de usuario para el login. Único por negocio, no globalmente
     * (constraint compuesta UNIQUE(negocio_id, username) a nivel de BD).
     */
    @Column(nullable = false, length = 50)
    private String username;

    /**
     * Contraseña del usuario (almacenada en hash).
     */
    @Column(nullable = false)
    private String password;

    /**
     * Indica si el usuario tiene permisos de jefe de cocina.
     * Los jefes de cocina pueden crear, editar y eliminar recetas.
     */
    @Column(nullable = false)
    private Boolean esJefeCocina = false;

    /**
     * Fecha y hora de registro del usuario en el sistema.
     */
    @Column(nullable = false)
    private LocalDateTime fechaRegistro;

    /**
     * Correo electrónico del usuario para notificaciones.
     */
    @Column(length = 100)
    private String email;


    /**
     * Constructor con parámetros para crear un usuario.
     *
     * @param username Nombre de usuario único
     * @param password Contraseña (debe ser hasheada antes de pasar)
     * @param esJefeCocina Indica si es jefe de cocina
     */
    public Usuario(String username, String password, Boolean esJefeCocina) {
        this.username = username;
        this.password = password;
        this.esJefeCocina = esJefeCocina;
        this.fechaRegistro = LocalDateTime.now();
    }
}